### PR TITLE
Pull request for sessions 3/4

### PR DESCRIPTION
**Pull request by ghost**
Please forgive the late pull request. I had surgery around the 2nd class was out (on disability) for a week and am still catching up. Thanks.

### CODE REVIEW

File: students/timm67/session03/slice_demo.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+def slice_exchange_first_last(seq):
+    firstElement = seq[0]
+    lastElement = seq[-1]
+    seq[0] = lastElement
+    seq[len(seq)-1] = firstElement
+    return seq
+
+def slice_rm_ev_other(seq):
+    sliceObj = slice(0, -1, 2)
```
Comment (hosungs) [inline students/timm67/session03/slice_demo.py:16]: Interesting... Never knew this type!

Why not just `return seq[::2]`?

File: students/timm67/session03/slice_demo.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    return front_back_slice[sliceObj]
+    
+def slice_reverse_seq(seq):
+    # NOTE: In Python 3, use two forward slashes for floor (integer) division
+    for x in range(0, (len(seq)//2)):
+        temp = seq[((len(seq) - 1) - x)]
+        seq[((len(seq) - 1) - x)] = seq[x]
+        seq[x] = temp
+    return seq
```
Comment (hosungs) [inline students/timm67/session03/slice_demo.py:35]: Again, desired code is simply `return seq[::-1]`

File: students/timm67/session03/slice_demo.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+def slice_last_third_first_third_mid_third_new_order(seq):
+    return seq
```
Comment (hosungs) [inline students/timm67/session03/slice_demo.py:38]: Not done...

File: students/timm67/session04/mailroom.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        donors[donor_name].append(donor_amt)
+    else:
+        # create a new donor (key)
+        print("{0} NOT found in dictionary, creating...".format(donor_name))
+        donors[donor_name] = [donor_amt]
+    print(donors[donor_name])
+    a = 0.0
+    donations = donors[donor_name]
+    for x in donations:
+        a += x
```
Comment (hosungs) [inline students/timm67/session04/mailroom.py:48]: Minor/FYI: `a = sum(donations)`

**Review by hosungs — COMMENTED** [review]
Thanks Tim for your PR. I see that this PR includes previous assignments as well as mailroom part 2 (for week 4), so I'm not sure how these are all graded on Canvas. I'll leave this PR open for the remaining week 4 assignment (Kata Fourteen). Also please work with TA regarding Canvas grading, if you haven't made Canvas submissions (if you did, please ignore).

File: students/timm67/session03/slice_demo.py (hunk 2 above)
Comment (ghost) [reply students/timm67/session03/slice_demo.py:35]: Learned this a few days ago; I guess it is an 'extended' slice operator

File: students/timm67/session03/slice_demo.py (hunk 3 above)
Comment (ghost) [reply students/timm67/session03/slice_demo.py:38]: Will review and add -- I was not sure about the instructions and will ask in class.

File: students/timm67/session04/mailroom.py (hunk 4 above)
Comment (ghost) [reply students/timm67/session04/mailroom.py:48]: Thanks for the hint.

File: students/timm67/session04/mailroom_part2.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+# 'amounts' remains a list of donation values 
+# 
+
+donors = [
+    { 'fname' : 'Jim',  'lname' : 'Tillson',  'amounts' : [5.00, 20.00]},
+    { 'fname' : 'Barb', 'lname' : 'Langley',  'amounts' : [10.00, 20.00, 100.00]},
+    { 'fname' : 'Jen',  'lname' : 'Garfield', 'amounts' : [10.00, 20.00, 5.00]},
+    { 'fname' : 'Rex',  'lname' : 'Miller',   'amounts' : [20.00, 20.00, 5.00]},
+    { 'fname' : 'Tony', 'lname' : 'Blake',    'amounts' : [20.00, 20.00, 10.00]},
+]
```
Comment (hosungs) [inline students/timm67/session04/mailroom_part2.py:23]: I think the intended dict structure is something like below:
```
donors = { 'Jim Tillson': [5, 20], 'Barb Langley': [10, 20, 100], ... }
```
Or, if you want to keep first names and last names separately, something like this:
```
donors = { ('Jim', 'Tillson'): [5, 20], ('Barb', 'Langley'): [10, 20, 100], ... }
```
The key is to be able to use `if (donor_fname, donor_lname) in donors:`, without having to search for a specific entry in the list. Let me know if this doesn't make sense... Not sure if I should require this change, but definitely this is something you will want to consider.

File: students/timm67/session04/mailroom_part2.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    
+    for donor in donors:
+        if donor['fname'] == donor_fname and donor['lname'] == donor_lname:
+            print("{0}, {1} FOUND in dictionary".format(donor_lname, donor_fname))
+            donor['amounts'].append(donor_amt)
+            donor_found = True
+            break
+
+    # if we got to here, donor is not in dictionary, so create a new donor
+    if donor_found == False:
```
Comment (hosungs) [inline students/timm67/session04/mailroom_part2.py:48]: Minor: Python's `else:` for loops (`for` or `while`) can be useful here. No need to use/set this flag variable at all.

File: students/timm67/session04/mailroom_part2.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        print("{0}, {1} NOT found in dictionary, creating...".format(donor_lname, donor_fname))
+        new_donor = { 'fname' : str(donor_fname), 'lname' : str(donor_lname), [].append(donor_amt) }
+        donors.append(new_donor)
+
+    a = 0.0
+    donations = donors[donor_name]
+    for x in donations:
+        a += x
```
Comment (hosungs) [inline students/timm67/session04/mailroom_part2.py:56]: Minor: `a = sum(donations)`

File: students/timm67/session04/mailroom_part2.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    fmtline0 = "\n\nDear {0},\n"
+    fmtline1 = "Many thanks for your recent donation of {0:6.2f}. With this donation, your cumulative total is {1:6.2f}\n"
+    fmtline2 = "Thanks for your donation. \n\n"
+    print(fmtline0.format(donor_name))
+    print(fmtline1.format(donor_amt, a))
+    print(fmtline2)
```
Comment (hosungs) [inline students/timm67/session04/mailroom_part2.py:62]: I believe in mailroom part 2, we are supposed to write this letter to a file...?

File: students/timm67/session04/mailroom_part2.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    print("Mailroom Tasks")
+    print("[1] Send a Thank you")
+    print("[2] Create a report")
+    print("[9] Exit Mailroom")
+    pass
+
+def main():
+
+    exitMail = False
+    while(exitMail == False):
```
Comment (hosungs) [inline students/timm67/session04/mailroom_part2.py:88]: Minor: `while not exitMail:`

File: students/timm67/session04/mailroom_part2.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            exitMail = True
+        elif (response == 5):
+            pass
+        elif (response == 4):
+            pass
+        elif (response == 3):
+            pass
+        elif (response == 2):
+            create_report_task()            
+            pass
```
Comment (hosungs) [inline students/timm67/session04/mailroom_part2.py:106]: Unnecessary `pass`, here and many other places

**Review by hosungs — COMMENTED** [review]
Left some more comments. Leaving this PR open, as there are still work in progress.

**Comment by ghost** [comment]
Thank you for the help, Hosung. Actually, I had the dictionary close to
what you wanted in part 1. I think putting the name into the tuple is the
best way.

Yes, I have to finish the file writing part, but was trying to get
dictionary structure right beforehand.

'pass' is kind of like a # TODO for me. I use it where something should go.
Yes, perhaps a bit too frequently.

I will probably need your help on the Trigrams once I finish MB part 2.

Thanks again...

Tim

**Comment by hosungs** [comment]
I'll have to merge/close this PR for the next assignment. I'll be OK if you include your session 4 work in your session 5 PR, so no worries about that. Just keep working on all assignments (if any previous one wasn't finished yet). Thanks.

**Pull request merged**